include/eve/detail/skeleton.hpp

```diff
@@ -188,37 +207,43 @@ namespace eve::detail
   template<typename Func, typename... Ts>
   EVE_FORCEINLINE auto aggregate(Func &&f, Ts &&... ts)
   {
-    using wide_t = typename wide_result<Func, Ts...>::type;
+    using half_t = decltype(EVE_FWD(f)(lower(EVE_FWD(ts))...));
 
-    // Full sliceable case
-    if constexpr( is_fully_sliceable<wide_t,std::decay_t<Ts>...>() )
+    if constexpr(kumi::product_type<half_t>)
     {
-      wide_t that;
-
-      // If everything can be sliced in equal sub-parts, we save compile-time and
-      // simplify debugging by iterating over all smallest sub-parts
-      detail::apply<replication<wide_t>()>
-      ( [&]<typename... I>(I const&...)
-        {
-          ( ( aggregate_step<I::value>::perform ( EVE_FWD(f)
-                                                , that.storage()
-                                                , EVE_FWD(ts)...
-                                                )
-            ),...
-          );
-        }
-      );
-
-      return that;
+      return tuple_combiner( EVE_FWD(f)(lower(EVE_FWD(ts))...), EVE_FWD(f)(upper(EVE_FWD(ts))...));
     }
-    // Recursive case
     else
     {
-      // We end up there if we have a difference of # of replications inside
-      // This happens in conversions context or when we call a function on a
-      // AVX/AVX2 type with no implementation.
-      return wide_t{EVE_FWD(f)(lower(EVE_FWD(ts))...),
-                    EVE_FWD(f)(upper(EVE_FWD(ts))...)};
+      using wide_t = typename half_t::template rescale<typename half_t::cardinal_type::combined_type>;
+
+      // Full sliceable case
+      if constexpr( is_fully_sliceable<wide_t,std::decay_t<Ts>...>() )
```

Comment:
I think this tries to handle different cardinals.

the number of elements in all wides has to be the same. Otherewise it's not element wise.

UPD: or is it trying to do something else? I don't get it. All of our aggrgation is arrays of 2. What's with indexes?

Reply:
This is an old piece of code that days that if we have aggregate of more than 2 pieces and all inputs are all aggregate, we walk down the pieces array directly instead of going recursive.

This was an old fix we designed to eas aggregation compile time.

Reply:
that's not a thing anymore though, is it?

Reply:
See https://github.com/jfalcou/eve/pull/1756#discussion_r1527145937